creator-node/src/fileManager.js

```diff
@@ -95,17 +96,55 @@ async function saveFileToIPFSFromFS (req, srcPath, fileType, sourceFile, transac
 }
 
 /**
- * Save file to disk given IPFS multihash, and ensure availability.
- * @notice This will only work for non-dir files
- * Steps:
- *  - If file already stored on disk, return immediately and store to disk.
- *  - If file not already stored, fetch from IPFS and store to disk.
- *    - If multihash available on local ipfs node, retrieve file.
- *    - If multihash not available locally, fetch file from IPFS.
- *  - If file is not available via IPFS try other cnode gateways for user's replica set.
- *  - Add file to local ipfs node if not already there.
+ * Given a CID, saves the file to disk. Steps to achieve that:
+ * 1. do the prep work to save the file to the local file system including
+ * creating directories, changing IPFS gateway urls before calling _saveFileForMultihash
+ * 2. attempt to fetch the CID from a variety of sources
+ * 3. throws error if failure, couldn't find the file or file contents don't match CID,
+ * returns expectedStoragePath if successful
+ * @param {Object} req request object
+ * @param {String} multihash IPFS cid
+ * @param {String} expectedStoragePath file system path similar to `/file_storage/Qm1`
+ *                  for non dir files and `/file_storage/Qmdir/Qm2` for dir files
+ * @param {Array} gatewaysToTry List of gateway endpoints to try
  */
-async function saveFileForMultihash (req, multihash, expectedStoragePath, gatewaysToTry = []) {
+async function saveFileForMultihash (req, multihash, expectedStoragePath, gatewaysToTry) {
+  const storagePath = req.app.get('storagePath') // should be like `/file_storage'
+  const filePath = expectedStoragePath.replace(storagePath, '') // should be `/Qm1/Qm2` for dir and `/Qm1` for non-dir
+
+  // will be modified to directory compatible route later if directory
+  let gatewayUrlsMapped = gatewaysToTry.map(endpoint => `${endpoint.replace(/\/$/, '')}/ipfs/`)
+
+  // Check if the file we are trying to copy is in a directory and if so, create the directory first
+  // E.g if the expectedStoragePath is /file_storage/QmABC/Qm123, we check if filePath contains more than two parts
+  // NOTE - `/QmABC/Qm123 splits into ['', 'QmABC', 'Qm123']. The beginning empty quote is because of the initial slash
+  // eg. `/QmABC/Qm123` contains more than 2 parts and if so, it's a directory
+  // eg. `/QmABC` splits into ['', 'QmABC'] so it's exactly two parts so it's a non-dir file
+  const splitPath = filePath.split('/')
+  if (splitPath.length > 2) {
```

Comment:
should we make this an explicit check for length == 3 and error if > 3? or maybe this is fine, idk

Reply:
i'm not tied to one or the other. do you have a strong preference for this?

Reply:
length is 2 for top level files and > 2 for dirs so thats why this is the way it is

Reply:
eh i'm not sure either, its fine

Reply:
this is _fine_ imo but kind of wondering is there no better way to do this? just one idea

```
> y = path.dirname('/file_storage/QmABC')
'/file_storage'
> y = path.dirname('/file_storage/QmABC/Qm123')
'/file_storage/QmABC'
```

^ this would tell us based on the returned value, if it's 'file_storage' then root, otherwise it is a dir

Reply:
@hareeshnagaraj we could use path but it's still kinda weird cause we need to get the the middle cid so we need to split no matter what. basically the dir check with path looks like this
```js
const parsedStoragePath = path.parse(expectedStoragePath).dir
// check that the dir property. doesn't equal storagePath (`/file_storage`). if it's equal it's a non-dir file`
if (parsedStoragePath.dir !== storagePath && parsedStoragePath.includes(storagePath)) {
  ...this is a directory
}
```